File: BSIGeneralAffairDAL/DALEmployee.vb
Imports System.Data
Imports System.Data.SqlClient
Imports BSIGeneralAffairBO
Imports GeneralAffairInterface

Public Class DALEmployee
    Implements IEmployee

    Public database As DatabaseConfiguration = New DatabaseConfiguration()
    Private dataRead As SqlDataReader
    Public cmd As SqlCommand

    Public Function getAll() As List(Of Employee) Implements IEmployee.getAll
        Dim Employeis As New List(Of Employee)
        Try
            Dim strSP = "[HumanResource].[USP_ListEmployee]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            database.conn.Open()
            dataRead = cmd.ExecuteReader()
            If dataRead.HasRows Then
                While dataRead.Read
                    Dim employee As New Employee
                    Dim user As New User
                    employee.EmployeeIDNumber = dataRead("EmployeeID").ToString()
                    employee.User = New User()
                    employee.User.UserFullName = dataRead("Fullname").ToString()
                    employee.EmployeePosition = dataRead("Position").ToString()
                    employee.EmployeeGender = dataRead("Gender").ToString()
                    employee.Department = New Departement()
                    employee.Department.DepartementName = dataRead("Department").ToString()
                    employee.Office = New Office()
                    employee.Office.OfficeName = dataRead("Office").ToString()
                    employee.EmployeeHireDate = dataRead("HireDate").ToString()
                    Employeis.Add(employee)
                End While
            End If
            dataRead.Close()

            Return Employeis
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function create(obj As Employee) As Integer Implements IEmployee.create
        Try
            Dim strSP = "[HumanResource].[USP_CreateEmployee]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure

            cmd.Parameters.AddWithValue("@Firstname", obj.User.UserFirstName)
            cmd.Parameters.AddWithValue("@LastName", obj.User.UserLastName)
            cmd.Parameters.AddWithValue("@EmployeeNumber", obj.EmployeeIDNumber)
            cmd.Parameters.AddWithValue("@DepartementID", obj.Department.DepartementID)
            cmd.Parameters.AddWithValue("@OfficeID", obj.Office.OfficeID)
            cmd.Parameters.AddWithValue("@EmployeePositionLevel", obj.EmployeePositionLevel)
            cmd.Parameters.AddWithValue("@EmployeeJobTitle", obj.EmployeeJobTitle)
            cmd.Parameters.AddWithValue("@EmployeeGender", obj.EmployeeGender)
            cmd.Parameters.AddWithValue("@EmployeeMaritalStatus", obj.EmployeeMaritalStatus)
            cmd.Parameters.AddWithValue("@EmployeeBirthDate", obj.EmployeeBirthDate)
            cmd.Parameters.AddWithValue("@EmployeeHireDate", obj.EmployeeHireDate)
            cmd.Parameters.AddWithValue("@EmployeeSalary", obj.EmployeeSalary)
            cmd.Parameters.AddWithValue("@EmployeeType", obj.EmployeeType)
            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
            If result = 1 Then
                Throw New ArgumentException("Employee not created")
            End If
            Return result
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function
End Class

File: BSIGeneralAffairDAL-C/DALEmployee.cs
using BSIGeneralAffairBO_C;
using BSIGeneralAffairDAL_C.Interfaces;
using Dapper;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SqlClient;
using System.Text;
using static Dapper.SqlMapper;

namespace BSIGeneralAffairDAL_C
{
    public class DALEmployee : IEmployeeDAL

    {
        private string GetConnectionString()
        {
            return Helper.GetConnectionString();
            //return @"Data Source=ACTUAL;Initial Catalog=LatihanDb;Integrated Security=True;TrustServerCertificate=True";
            //return ConfigurationManager.ConnectionStrings["MyDbConnectionString"].ConnectionString;
        }

        public void Create(Employee employee)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = "[HumanResource].[USP_CreateEmployee]";
                var param = new {
                    Firstname = employee.User.UserFirstName,
                    LastName = employee.User.UserLastName,
                    EmployeeNumber = employee.EmployeeIDNumber,
                    DepartementID = employee.Department.DepartementID,
                    OfficeID = employee.Office.OfficeID,
                    EmployeePositionLevel = employee.EmployeePositionLevel,
                    EmployeeJobTitle = employee.EmployeeJobTitle,
                    EmployeeGender = employee.EmployeeGender,
                    EmployeeMaritalStatus = employee.EmployeeMaritalStatus,
                    EmployeeBirthDate = employee.EmployeeBirthDate,
                    EmployeeHireDate = employee.EmployeeHireDate,
                    EmployeeType = employee.EmployeeType,
                    EmployeeSalary = employee.EmployeeSalary };
                try
                {
                    int result = conn.Execute(strSql, param, commandType: System.Data.CommandType.StoredProcedure);
                    if (result == 1)
                    {
                        throw new ArgumentException("Insert data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new ArgumentException($"{sqlEx.InnerException.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new ArgumentException(ex.Message);
                }
            }
        }

        public IEnumerable<EmployeeList> GetAll()
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                //var strSql = @"select * from Categories order by CategoryName";
                var strSql = @"[HumanResource].[USP_ListEmployee]";

                var results = conn.Query<EmployeeList>(strSql, commandType: System.Data.CommandType.StoredProcedure);
                return results;
            }
        }

        public void Update(Employee employeeUpdate)
        {
            throw new NotImplementedException();
        }

        public EmployeeList GetById(int employeeNumberid)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"[HumanResource].[USP_ListEmployeeByID]";
                var param = new { EmployeeIDNumber = employeeNumberid };
                var result = conn.QuerySingleOrDefault<EmployeeList>(strSql, param);
                return result;
            }
        }

        public IEnumerable<EmployeeList> GetByName(string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"[HumanResource].[USP_ListEmployeeByName]";
                var param = new { Fullname = name };
                var result = conn.Query<EmployeeList>(strSql, param);
                return result;
            }
        }

        public int GetCountEmployee(string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"[HumanResource].[USP_CountEmployeeByName]";
                var param = new { Fullname = $"%{name}%" };
                var result = Convert.ToInt32(conn.ExecuteScalar(strSql, param));
                return result;
            }
        }

        public IEnumerable<EmployeeList> GetWithPaging(int pageNumber, int pageSize, string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"[HumanResource].[USP_PaginationEmployeeByName]";
                var param = new { Fullname = $"%{name}%", Offset = (pageNumber - 1) * pageSize, PageSize = pageSize };
                var results = conn.Query<EmployeeList>(strSql, param);
                return results;
            }
        }
    }
}
